pp_netlib/functions.py

```diff
@@ -210,3 +286,420 @@ def save_graph(graph, backend, outdir, file_name, file_format):
+           Minimum spanning tree (as graph-tool graph)
+    """
+    #
+    # Create MST
+    #
+
+    mst_network = graph
+
+    # Find seed nodes as those with greatest outdegree in each component
+    num_components = 1
+    seed_vertices = set()
+    
+    mst_df = cugraph.components.connectivity.connected_components(mst_network)
+    num_components_idx = mst_df["labels"].max()
+    num_components = mst_df.iloc[num_components_idx]["labels"]
+    if num_components > 1:
+        mst_df["degree"] = mst_network.in_degree()["degree"]
+        # idxmax only returns first occurrence of maximum so should maintain
+        # MST - check cuDF implementation is the same
+        max_indices = mst_df.groupby(["labels"])["degree"].idxmax()
+        seed_vertices = mst_df.iloc[max_indices]["vertex"]
+
+    # If multiple components, add distances between seed nodes
+    if num_components > 1:
+
+        # Extract edges and maximum edge length - as DF for cugraph
+        # list of tuples for graph-tool
+        
+        # With cugraph the MST is already calculated
+        # so no extra edges can be retrieved from the graph
+        G_df = graph.view_edge_list()
+        max_weight = G_df["weights"].max()
+        first_seed = seed_vertices.iloc[0]
+        G_seed_link_df = cudf.DataFrame()
+        G_seed_link_df["dst"] = seed_vertices.iloc[1:seed_vertices.size]
+        G_seed_link_df["src"] = seed_vertices.iloc[0]
+        G_seed_link_df["weights"] = seed_vertices.iloc[0]
+        G_df = G_df.append(G_seed_link_df)
+
+    # Construct graph
+    mst_network = cugraph.Graph()
+    G_df.rename(columns={"src": "source","dst": "destination"}, inplace=True)
+    mst_network.from_cudf_edgelist(G_df, edge_attr="weights", renumber=False)
+
+    sys.stderr.write("Completed calculation of minimum-spanning tree with CU.\n")
+
+    return mst_network
+
+def generate_mst_network(graph, backend):
+    unweighted_error_msg = RuntimeError("MST passed unweighted graph, weighted tree required.")
+
+    if backend == "GT":
+        if "weight" not in graph.edge_properties:
+            raise unweighted_error_msg
+        else:
+            mst_network = gt_generate_mst(graph)
+
+    elif backend == "NX":
+        if "weight" not in list(list(graph.edges(data=True))[0][-1].keys()): ## https://stackoverflow.com/questions/63610396/how-to-get-the-list-of-edge-attributes-of-a-networkx-graph
+            raise unweighted_error_msg
+        else:
+            mst_network = nx_generate_mst(graph) ##TODO
+
+    elif backend == "CU":
+        if not graph.is_weighted():
+            raise unweighted_error_msg
+        else:
+            mst_network = cu_generate_mst(graph)
+
+    return mst_network
+
+def get_gt_clusters(graph):
+    import graph_tool.all as gt
+
+    vertex_labels = list(str(graph.vp["id"][v]) for v in graph.vertices())
+
+    # get a sorted list of component assignments
+    component_assignments, component_frequencies = gt.label_components(graph)
```

Comment:
in the cleanup branch i've been working on, i've made it so that prepare_graph calls this function. Should I leave this as is for now or bring those changes here?